projects/capybara/htm/htm_network_runner.py

```diff
@@ -0,0 +1,284 @@
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+import argparse
+import csv
+import json
+import logging
+import numpy as np
+import os
+import time
+import yaml
+
+from htm_network import BaseNetwork
+
+logging.basicConfig()
+_LOGGER = logging.getLogger('NetworkRunner')
+_LOGGER.setLevel(logging.INFO)
+
+
+
+def _newTrace():
+  return {
+    't': [],
+    'scalarValue': [],
+    'encoderOutput': [],
+    'spActiveColumns': [],
+    'tmActiveCells': [],
+    'tmPredictiveCells': [],
+    'tmPredictedActiveCells': [],
+    'rawAnomalyScore': [],
+    'label': []
+  }
+
+
+
+class NetworkRunner(object):
+  def __init__(self, batchSize, inputMin, inputMax, runSanity):
+
+    self.batchSize = batchSize
+    self.network = BaseNetwork(inputMin=inputMin,
+                               inputMax=inputMax,
+                               runSanity=runSanity)
+    self.network.initialize()
+    self.traceNZ = _newTrace()
+
+    # Keep track of the min and max value observed as we stream the data.
+    self.observedMin = None
+    self.observedMax = None
+
+    # Keep track of repeating encodings
+    self.lastEncoding = None
+    self.encodingRepetitionCounter = 0
+    self.encodingRepetitionThreshold = 2
+
+    # Keep track of the processing time
+    self.startTime = time.time()
+
+
+  def _updateTrace(self, t, scalarValue, label, encoderOutput,
+                   spActiveColumns, tmActiveCells,
+                   tmPredictiveCells, tmPredictedActiveCells, rawAnomalyScore):
+
+    self.traceNZ['t'].append(t)
+    self.traceNZ['scalarValue'].append(scalarValue)
+    self.traceNZ['label'].append(label)
+    self.traceNZ['encoderOutput'].append(encoderOutput)
+    self.traceNZ['spActiveColumns'].append(spActiveColumns)
+    self.traceNZ['tmActiveCells'].append(tmActiveCells)
+    self.traceNZ['tmPredictiveCells'].append(tmPredictiveCells)
+    self.traceNZ['tmPredictedActiveCells'].append(tmPredictedActiveCells)
+    self.traceNZ['rawAnomalyScore'].append(rawAnomalyScore)
+
+
+  def _writeTraceBatch(self, traceWriter):
+    numPoints = len(self.traceNZ['t'])
+    for i in range(numPoints):
+      row = []
+      for traceName in self.traceNZ.keys():
+        trace = self.traceNZ[traceName]
+        if trace:
+          if type(trace[i]) == np.ndarray:
+            trace[i] = trace[i].tolist()
+          row.append(json.dumps(trace[i]))
+        else:
+          row.append(None)
+      traceWriter.writerow(row)
+
+
+  def _setObservedMinMax(self, scalarValue):
+    if self.observedMin is None:
+      self.observedMin = scalarValue
+    elif self.observedMin > scalarValue:
+      self.observedMin = scalarValue
+
+    if self.observedMax is None:
+      self.observedMax = scalarValue
+    elif self.observedMax < scalarValue:
+      self.observedMax = scalarValue
+
+
+  def _isEncodingChanging(self, encoding):
```

Comment:
@oxtopus Thanks for the feedback, again. I made some changes here: https://github.com/numenta/nupic.research/pull/709/commits/418acbc77611bc4eb46dcd5cf14cccbb0f860dec